### DISCUSSION TITLE
Is it possible to configure `MiniStatusline.config.diagnostic_levels`?

### DISCUSSION THREAD

**Discussion by HoldenLucas** [nvim-mini/mini.nvim#41] [Q&A] [answered]
https://github.com/echasnovski/mini.nvim/blob/c77359a61d5f0f27603bbc61074e3890a38506c8/lua/mini/statusline.lua#L470

This line makes me think "no", but I started using `mini` today and haven't poked around very much. Ultimately I'd like to have separate sections for `Error`, `Warning`, and `Hint`, each tied to their respective highlights and shown only if their respective diagnostic lists are populated. Should I just define my own sections?

Also, thank you so so much for letting me replace ~100 lines of `cmp` config with a one liner. `mini` has already replaced 30% of my plugins.

**Comment by echasnovski** [comment, marked as answer]
Not at the moment, no. Not really because of that line (which is actually redundant, so thank you for pointing this out :) ).

I'll think for a while about adding this here somehow, as it seems like a reasonable usage. But currently I can't see a good way of doing this that is better than custom minimal section function. Here is my version for Neovim 0.6+ (add it before call to `require('mini.statusline').setup()`):
```lua
local diagnostic_level = function(level, prefix)
  if MiniStatusline.is_truncated(75) then
    return ''
  end

  local n = #vim.diagnostic.get(0, { severity = vim.diagnostic.severity[level] })
  return (n == 0) and '' or ('%s%s'):format(prefix, n)
end
```

After that it can be used something like this:
```lua
require('mini.statusline').setup({
  content = {
    active = function()
      -- ...
      local errors = diagnostic_level('ERROR', 'E')
      local warns  = diagnostic_level('WARN', 'W')
      local hints  = diagnostic_level('HINT', 'H')
      -- ...

      return MiniStatusline.combine_groups({
        -- ...
        { hl = 'DiagnosticError', strings = { errors } },
        { hl = 'DiagnosticWarn',  strings = { warns } },
        { hl = 'DiagnosticHint',  strings = { hints } },
        -- ...
    end,
  },
})
```

**Edit**: Oh, and thank you for kind words. `nvim-cmp` has **much, much** richer functionality though, and I'd suggest using 'mini.completion' only if you fully aware that it will be enough for you (because I most likely won't implement any new features for it, only bug fixes).

**Comment by verygitmuchhub** [comment]
Hey there,

I'm pretty much interested in switching to mini.statusline, but whatever I do, .. I don't get how to express the configuration in init.lua to support custom sections.

I then found this example code here and was like: FINALLY some example to follow! \o/ Just to find out that it still does not answer this one question that bugs me with like every mini plugin I'm using:

How do I customize mini things? So, I've read all the docs, and I understand everything, up to that point when I want to write such configuration things in my init.lua.

I have the following (Lazy) configuration (that tells me):

> Lua Diagnostics.: Undefined global `MinitStatusline`. [undefined-global]

```lua
require('lazy').setup({
  { 'echasnovski/mini.statusline',
    config = function()
      vim.api.nvim_set_hl(0, 'MiniStatuslineModeCommand', { ... })
      -- ...

      require('mini.statusline').setup({
        content = {
          active = function()
            local mode, mode_hl = Ministatusline.section_mode({ trunc_width = 120 }) -- Undefined global Ministatusline here
            -- ...
          end,
        }
      })
    end,
  },
})
```

Please please please someone enlighten me. Thanks!

**Reply by echasnovski** [reply]
The code regarding 'mini.statusline' looks correct. The warning can be ignored: `require('mini.statusline').setup()` will create `_G.MiniStatusline` global variable which can be used instead of `require('mini.statusline')`, but LuaLS doesn't understand that.

If you don't like that, you can do something like this:

```lua
      local statusline = require('mini.statusline')
      statusline.setup({
        content = {
          active = function()
            local mode, mode_hl = statusline.section_mode({ trunc_width = 120 })
            -- ...
          end,
        }
      })
```

Statusline structure is defined [via defining function for active and inactive windows](https://github.com/echasnovski/mini.nvim/blob/48f48e4b3f317e9da34ee7a01958b4c5018e2d34/doc/mini-statusline.txt#L9-L13). Each such function should return a string which is treated as a value of 'statusline' option (i.e. see `:h 'statusline'` for special placeholders it may use).

The definition is done by setting [`config.content.active` and/or `config.content.inactive`](https://github.com/echasnovski/mini.nvim/blob/48f48e4b3f317e9da34ee7a01958b4c5018e2d34/doc/mini-statusline.txt#L141-L148). The default value to use as a reference is described [here](https://github.com/echasnovski/mini.nvim/blob/48f48e4b3f317e9da34ee7a01958b4c5018e2d34/doc/mini-statusline.txt#L82).

**Reply by verygitmuchhub** [reply]
Thanks @echasnovski. I was pretty sure this cannot be too hard. At least I'm now able to control what's shown \o/ Let's see whether I can make my custom "trailing blank lines" and "mixed indent" functions part of this.
